aten/src/ATen/native/cuda/ReduceOpsKernel.cu

```diff
@@ -222,6 +224,229 @@ void argmin_kernel_cuda(TensorIterator& iter) {
+__device__ scalar_t binary_op_checking_nan(scalar_t lhs, scalar_t rhs, BinaryOperation binary_op) {
+  return (std::isnan(rhs) || (!std::isnan(lhs) && binary_op(rhs, lhs))) ? rhs : lhs;
+}
+/* Perform an inclusive scan along the innermost dimension of a tensor.
+ *
+ * - num_rows is the size of the flattened outer dimensions;
+ * - row_size is the size of the innermost dimension;
+ *
+ * The outer dimensions of the tensor are considered as a single dimension, i.e. the tensor is
+ * considered as having 'num_rows' rows of size 'row_size'.
+ * Each thread block processes one or more sets of contiguous rows (processing multiple rows
+ * per thread block is quicker than processing a single row, especially for short rows).
+ */
+template<typename scalar_t, int num_threads_x, int num_threads_y, class BinaryFunction>
+__global__ void tensor_kernel_scan_innermost_dim_with_indices(const scalar_t *self_, scalar_t *values_, int64_t *indices_,
+                                                int num_rows, int row_size,
+                                                scalar_t init, BinaryFunction binary_op) {
+  __shared__ scalar_t vbuf[num_threads_y][2 * num_threads_x];
+  __shared__ int64_t ibuf[num_threads_y][2 * num_threads_x];
+  scalar_t* row_buf = vbuf[threadIdx.y];
+  int64_t* row_idx_buf = ibuf[threadIdx.y];
+
+  for (int block_row = blockIdx.x * blockDim.y;
+       block_row < num_rows;
+       block_row += blockDim.y * gridDim.x) {
+    int row = block_row + threadIdx.y;
+    const scalar_t *row_self = self_ + row * row_size;
+    scalar_t *row_values = values_ + row * row_size;
+    int64_t *row_indices = indices_ + row * row_size;
+    scalar_t block_total = init;
+    int64_t block_idx_final = 0;
+    // Perform scan on one block at a time, keeping track of the total value of
+    // all blocks processed so far.
+    for (int block_col = 0; block_col < row_size; block_col += 2 * num_threads_x) {
+      // Load data into shared memory (two values per thread).
+      int col1 = block_col + threadIdx.x;
+      int col2 = block_col + num_threads_x + threadIdx.x;
+      if (row < num_rows) {
+        if (col1 < row_size) {
+          row_buf[threadIdx.x] = row_self[col1];
+          row_idx_buf[threadIdx.x] = col1;
+        } else {
+          row_buf[threadIdx.x] = init;
+          // No need to set the index here as the value in init will never be selected
+        }
+
+        if (col2 < row_size) {
+          row_buf[num_threads_x + threadIdx.x] = row_self[col2];
+          row_idx_buf[num_threads_x + threadIdx.x] = col2;
+        } else {
+          row_buf[num_threads_x + threadIdx.x] = init;
+          // No need to set the index here as the value in init will never be selected
+        }
+
+        // Add the total value of all previous blocks to the first value of this block.
+        if (threadIdx.x == 0) {
+          auto out = binary_op_checking_nan(block_total, row_buf[0], binary_op);
+          if(out == block_total) {
```

Comment:
suppose your `block_total` from previous iteration was `nan`. Then your `row_buf[0]` will correctly be set to `nan`, but your `row_idx_buf[0]` won't be set to your previous index corresponding to `nan` element, because the comparison above will be false (`nan ==nan` evaluates to `false`)